On csv upload allow column promotion from boolean to int / float

wotbrew opened the pull request:
Closes https://github.com/metabase/metabase/issues/48662

### Description

Currently, a .csv column containing int values 0 and 1 will be mapped to a boolean column.

If a user then uploads a second .csv (append or replace) containing integer values outside of 0 or 1 they receive an error on import (e.g '2' is not recognizable as a boolean).

This PR permits boolean columns to be migrated to integer or float columns when encountering input outside their range

Quirks/observations on interaction between this and existing coercion / narrowing / widening behaviour in regards to floats:
 
- boolean columns can now promote to float columns
- boolean representable floats 1.0, 0.0 will widen the column to a float
   - one could (mistakenly IMO) expect a kind of transitivity in regards to the `*float-or-int*` + int coercion behaviour, maintaining a boolean column.
- integer representable floats such as '2.0' and '42.0' will also widen the column to a float
   - this seems very reasonable, just pointing it out

Changes confirmed to work for H2, MySQL, Postgres, and Clickhouse cloud. 

- Clickhouse cloud was tested manually and by running the new tests against a trial cloud instance
- Redshift does not support the new promotions, should yield the same 'X is not of type Y' style error message
- Postgres has some additional mechanisms via `USING` to get boolean to bigint and float to work, as postgres does not want to `CAST` those columns automatically

### How to verify

1. Upload a .csv `col\n0`, get a boolean column with the value `false`.
2. Append or replace with a .csv `col\n2`, column should be promoted to an int with values `0,2`.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

wotbrew committed:
On csv upload allow column promotion from boolean to int / float

crisptrutski commented on src/metabase/upload/types.cljc:
```diff
@@ -251,7 +257,11 @@
                 (if (str/blank? trimmed)
                   current-type
-                  (->> (cons current-type (ancestors h current-type))
+                  (->> (concat
+                         [current-type]
+                         (if-some [abstraction (widening-abstraction current-type)]
+                           (cons abstraction (ancestors h abstraction))
+                           (ancestors h current-type)))
```
nit: tastes vary, and i'm sure we'll both converge on shared aesthetics eventually, but the mixture of `concat`, a single element vector, and a complex nested form just rubs me the wrong way both syntactically and in execution.

What do you think about something like this instead?

```clojure
(let [t (concrete->abstract current-type)]
  (list* current-type t (ancestors h t)))
```

We can just filter out the nil. Actually on second thought, I'm not sure we need to prepend with current-type. I *think* we can assume that `concrete->abstract` will always project to a descendant[^1], in which case the current-type will be traversed as an ancestor, and we actually want it to be tried later for more complex type hierarchies. 

So all we would need is:

```clojure
(let [t (concrete->abstract current-type current-type)]
  (cons t (ancestors h t)))
```

Yeah, I like short names when in a tight scope. You could even use `as->` but I prefer the let here.

[^1]: Since I think that `abstract->concrete` must always point to a parent for things to make sense. It's been a while, and I haven't fully reloaded the context.

wotbrew replied:
Yep, I had the same thought about dropping current-type. Pushed a much terser form. (See comment about widening abstraction though, this might change this anyway).

crisptrutski reviewed (commented):
The change in behavior looks correct to me. Once we accept that we will alter the schema, we might as well relax all the way to the same type we'd use for uploading all the rows from scratch.

Even though it's still draft, I've left some provisional nits, but I'd prefer to discuss any you disagree with, rather than treating them as passing suggestions.

wotbrew commented on src/metabase/upload/types.cljc:
```diff
@@ -243,6 +246,9 @@
            (filter #((type->check %) trimmed))
            first))))
 
+(def ^:private widening-abstraction
```
I totally agree, did not get around to good docs (or names!) yet.

The main thing I am unsatisfied about is this map and its usage. It causes some 'interesting' new relaxation behaviours, such as this: 

```
(upload-types/column-types-from-rows
  {:number-separators ".,"}
  [nil]
  [["y"]
   ["n"]
   ["1"]
   ["2"]])
=> (:metabase.upload.types/int)
```
This is a natural consequence of treating the boolean column as 'maybe meaning ints'. In our case the column representation is lossy (e.g the original symbolic representation, the fact it was 'y' or 't' or '1' is lost), and so we must say something like this to get the behaviour to work. 

I am thinking it might be better to distinguish between the type of a column (which on first import is nil) and the exising 'current type' parameter, which is used for the reduction accumulator.

crisptrutski replied:
100% agree it should be a varchar

crisptrutski replied:
I'm surprised by this behavior though, "y" should be parsed as `boolean` value rather than `*boolean-int*`, and `int` is not an ancestor of that, and their lowest common ancestor is varchar 😕

wotbrew replied:
You've already seen - but fixed in latest. Will work on some tests for this though, think its worth having as it might have uh... 'surprised and delighted' somebody.

crisptrutski commented on src/metabase/upload/types.cljc:
```diff
@@ -263,11 +264,22 @@
       (vec (u/map-all relax value-types row)))))
 
+(def ^:private column-type->abstract
+  "Before a type has been concretized, values have more degrees of freedom for relaxation.
+
+  e.g. The type given for '1' would be *boolean-or-int*, which can relax into an int if a '2' is encountered.
+
+  If we are appending or replacing a previous .csv, we no longer know how the existing values were encoded.
+  Therefore, this mapping allows us to give a type the 'the benefit of the doubt'
+  and assume relaxations that might have been available given all possible encodings."
```
Var name and comment are 👨‍🍳 💋

crisptrutski commented:
LGTM!

wotbrew commented:
As expected, driver fails. Will pick it up again next week.

crisptrutski commented:
Seems like H2 and MySQL schema migration works already? Postgres looks like it'll be simple enough. Have some trepidation about Redshift.

A very important driver not covered by our CI is Clickhouse, as that's whats used for Cloud customers that buy attached storage. Clickhouse maintain that community driver themselves, but unfortunately this functionality is not tested in their CI either. What I've done in the past is run the tests manually myself. The upload tests so far have passed against their local development server version, but to be sure we should run it against a cloud instance as well.

crisptrutski commented:
An idea: force drivers to declare which migrations they support, and land this change with Clickhouse and potentially redshift disabling this relaxation for now.

wotbrew committed:
Fix incorrect arglist

wotbrew committed:
Add postgres support for bool -> int, float upload migrations

wotbrew committed:
Improve compatibility with old drivers

crisptrutski commented on src/metabase/driver.clj:
```diff
@@ -1162,11 +1162,43 @@
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
 
+(defmulti alter-upload-columns!
```
How about `alter-table-colums` rather? Seems plausible that this whole data editing project will leverage it as well, for non uploaded tables.

wotbrew replied:
Hmm, I did think this might come up. I do think your position is reasonable, but I also think my position (conservative naming / narrower scoping) was reasonable in general (trade-offs). In this instance I defer to your judgement as you have far more awareness of style and wider context! Renamed in https://github.com/metabase/metabase/pull/52607/commits/40ea124477127d5df1990312eabf2f410d4e00f4

wotbrew committed:
Rename alter-upload-columns to alter-table-columns

wotbrew committed:
Formatting, sparking joy

crisptrutski commented on src/metabase/upload/types.cljc:
```diff
@@ -85,7 +85,8 @@
   If we require a relaxation which is not allowlisted here, we will reject the corresponding file."
-  {::int #{::float}})
+  {::int     #{::float}
+   ::boolean #{::int, ::float}})
```
In terms of getting this PR merged without Clickhouse or Redshift support, we just need to get the error messages right, and I can imagine two alternatives:

1. Keep the errors as they are today (X is not of type Y)
2. Inform the user that the required type promotions are not supported.

New messages would need to craft the wording and get it translated, but this approach is probably simpler from a technical point of view. Specializing promotion rules per driver could be a bit brittle, as we need to make sure that `allowed-promotions`, `column-type->abstract` and `abstract->concrete` are all consistent.

Trying and failing promotions is not very practical, and it would also be nice if we could report on all the relevant fields rather than the first one we hit, so I think exceptions in general are out.

There's a related problem of whether our database supports a given type, which we handle via optional values in [this multimethod](https://github.com/ClickHouse/metabase-clickhouse-driver/blob/cb907c16d98aaf8718d08c22deac41df3536f9c6/src/metabase/driver/clickhouse.clj#L196). The fact that this isn't data we can introspect, and that we always fallback to a string, are two mistakes in that implementation though. The general idea could work here as well though.

What do you think?

wotbrew replied:
I propose fixing the tests to expect an error from redshift (as we currently do). This gets the PR merged fastest, and we are not strictly in any worse position, the PR is an improvement for other drivers.

If there is appetite to improve the error messages (which I am not against!), I suggest a new issue describing that problem - and to work on that in the full scope of this and other column alteration use cases with data editing.

crisptrutski replied:
I'm not comfortable exposing a raw database error / "something went wrong" to Redshift users, compared to the clear error they would receive currently.

wotbrew replied:
I think they receive the same error as if they tried to promote an int to a float? Sorry If I have misunderstood. AFAICT [any kind of alter-column](https://github.com/metabase/metabase/blob/master/modules/drivers/redshift/src/metabase/driver/redshift.clj#L546) for redshift throws that exception. 

I'm sorting the tests to check we get the same error for bool -> int, bool -> float for redshift.

crisptrutski replied:
Since we don't have these promotions defined yet, I'm talking about the "2 is not a boolean" types of error message.

wotbrew replied:
I've added driver/upload-promotion-allowlist so that redshift can behave as it did before. This seemed the most straightforward option to me. Perhaps there is something I'm missing, but we can discuss properly on monday. 

Notes:
- do we want redshift to use an empty allow list, so that int -> float also gets a 'nice' error message?
- this means the clickhouse driver needs to get patched to get the new promotions - I can submit a PR for this after merge

wotbrew replied:
I'm gonna commit the tests assuming a `{}` allowlist for redshift, as if it is 'better' to have the parser errors in general, why wouldn't we - this is for a new major release after all. LMK if you disagree

crisptrutski replied:
That solution works, but I'd like a bit of hammock time on this proposal since we may have to live with that interface for a while.

Two concerns that jump out at me are:

1. Disconnection between this definition and separate but coupled code for coercion.
2. Repetition for transitivity, and potential for mistakes there.

Not saying this direction is a no-go or that there's definitely something superior, I just haven't had the bandwidth to think it through.

wotbrew replied:
Transitivity test and notes added as discussed check the last couple of commits:

Test: 5d076a5adf73
Docstring: 66ba5e67a72b

crisptrutski replied:
Feels good to resolve this thread!

crisptrutski commented on src/metabase/driver/postgres.clj:
```diff
@@ -982,19 +982,48 @@
+  extra input from the user."
+  [column new-type old-type]
+  (condp = [new-type old-type]
+    [[:bigint] [:boolean]]
```
Cosmetic nit, but I think its more intuitive to order the pair like `[old new]`, as if they were key entries in a mapping of what was supported.

wotbrew replied:
Agree, good spot. https://github.com/metabase/metabase/pull/52607/commits/51c8a83a62761ab3c45937006a89b482f98ffa9e

crisptrutski commented on src/metabase/driver/postgres.clj:
```diff
@@ -982,19 +982,48 @@
+     (quote-identifier column) 1.0
+     :else 0.0]
+
+    nil))
```
Further to my other comment, I'm a bit nervous about these "open" semantics, where we assume all other pairs can be implicitly converted. What do you think about just enumerating them with a sentinel value like `true`?

wotbrew replied:
Checkout the new docstring [f0e7d05](https://github.com/metabase/metabase/pull/52607/commits/f0e7d05ebbfa0bac55cd8013d630c29654e71959)

Not sure if this is satisfactory or not for now, trying to be more transparent about the limitations and implicit nature here. Feel like this fn will change anyway

crisptrutski replied:
> if one takes the implementation to be transparent in terms of alter column limitations what we have here is fine

Where my thinking was going is that we could make this method the source of truth for what is supported

wotbrew committed:
Flip alter-column-using-hsql-expr arg order

wotbrew committed:
Docstring tweak alter-column-using-hsql-expr

wotbrew committed:
More symmetry with existing tests, redshift

wotbrew committed:
Add driver changelog note re `alter-table-columns!`

wotbrew committed:
Prefer ::driver over :default

wotbrew committed:
Allow drivers to specify the promo allowlist

wotbrew committed:
Prefer parser errors for redshift upload promos

wotbrew committed:
Ensure promo tests respect driver allowlist

wotbrew committed:
Make index-test/transitive available as a test util

wotbrew committed:
Add transitivity test

wotbrew committed:
Add allowlist note about transitivity requirement

crisptrutski commented on src/metabase/driver.clj:
```diff
@@ -1215,6 +1248,38 @@
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
 
+(defmulti upload-promotion-allowlist
```
Getting deep in the nits here, but a suggested rename:

```suggestion
(defmulti allowed-promotions
```

1. I can't see any reason we'd want to artificially limit which promotions are allowed on a driver-by-driver bases. The relaxation rules in CSV restrict us to everything that makes sense, and then each driver should just be best effort.
2. It's not a list.
3. Natural english word order. Highlights the noun.
4. Avoids jargon.

wotbrew replied:
821398f227d9

crisptrutski commented on src/metabase/driver.clj:
```diff
@@ -1215,6 +1248,38 @@
   :hierarchy #'hierarchy)
 
+(defmulti upload-promotion-allowlist
+  "Returns a mapping of which types a column can be implicitly relaxed to, based on the content of appended values.
```
Commenting on the next line to avoid interleaving on one thread.

Summary of our discussion around adding this multimethod, for the imaginary readers of the future:

Against:

1. There is brittle coupling with the set of values we defined coercion expressions for.
2. When adding 
3. We could derive this from a more data-centric expression of coercion rules.
4. More driver surface area. Our surface area is growing at an unsustainable rate.

Defense:

1. This is the simplest addition we could make to this PR to get it merged.
2. Very few drivers even implement uploads, that we know of, and we're going to update them all ourselves.
3. There's a bunch of other stuff we'll need to consolidate in this area of the drivers, so we can clean this up at the same time if it chafes us.
4. We could give this a default implementation that derives it from a data-structure of coercion rules, for a smooth deprecation.
5. The "brittle coupling" should be well covered by tests anyway. 
  - The lack of CI for CH notwithstanding.

That's all I remember.

crisptrutski commented on src/metabase/driver/postgres.clj:
```diff
@@ -1006,23 +1006,58 @@
+  It returns nil if no such expression has been defined for the pair of types. In this case, the caller should
+  generate the ALTER COLUMN statement without a USING."
+  [column old-type new-type]
+  (condp = [old-type new-type]
```
nit: I think this can just be `case` statement

wotbrew replied:
e39fe33326d7

crisptrutski commented on src/metabase/driver/sql_jdbc.clj:
```diff
@@ -206,10 +206,24 @@
+  [driver db-id table-name column-definitions & opts]
+  (let [deprecated-default-method     (get-method driver/alter-columns! :sql-jdbc)
+        deprecated-driver-method      (get-method driver/alter-columns! driver)
+        deprecated-method-specialised (not (identical? deprecated-default-method deprecated-driver-method))]
```
nit: prefer an `?` to end boolean variable.

Otherwise "specialized" here could be taken as an adjective, that's actually my first reading.

wotbrew replied:
f2acc2a69b3d

FWIW This is one of those personal style vs project style problems, so will keep an eye out for me doing this. Early in my clj journey I used to put ? on variables/kw flags like java isFoo, isBar, stopped doing that many years ago - reserving it only for predicates. Have to relearn the impulse to `?`. I do understand many projects prefer the ? on everything.

crisptrutski commented on src/metabase/driver/sql_jdbc/sync/interface.clj:
```diff
@@ -133,10 +133,34 @@
+  [driver table-name column-definitions & _opts]
+  (alter-columns-sql driver table-name column-definitions))
+
+;; default :sql-jdbc implementation kept here rather than on alter-table-columns-sql
+;; to maximize compatibility:
+;; a. get-method for superclass type calls still pick up this impl
+;; b. existing specialisation of sql-jdbc should remain preferred when alter-table-column-sql is called
+;; we can move this impl to alter-table-columns-sql when alter-columns-sql is deleted (v0.57+)
```
Nice idea

crisptrutski commented on test/metabase/upload_test.clj:
```diff
@@ -2205,28 +2205,64 @@
+(defn- external-type [t]
+  (keyword "metabase.upload" (name t)))
+
+(defn- promo-allowed? [allowlist col-type new-type]
```
This is always called with the same `allowlist`. Why not close over it?

crisptrutski replied:
Oh nevermind, reader error.

crisptrutski commented on src/metabase/upload.clj:
```diff
@@ -755,48 +763,58 @@
     ;; Ideally we would do all the filtering in the query, but this would not allow us to leverage mlv2.
     (model-persistence/invalidate! {:card_id [:in model-ids]})))
 
+(defn- internal-allowlist [m]
```
Two thoughts:

1. Name is a bit opaque, would rather name it functionally: `translate-type-keywords` or similar.
2. Do we need to translate between namespaces like this? 
 The [light-weight aliasing](https://clojure.atlassian.net/browse/CLJ-2123) added in 1.11 will make it a cinch.

wotbrew replied:
On 1. 20b3c0341e21

On 2.

I am aware of aliasing but I do not see how it helps in this instance. Drivers currently use the `metabase.upload` namespace, not `metabase.upload.types` (see `driver/upload-type->database-type`). I took this to mean there os a public API / internal distinction, hence the choice to preserve the distinction - and thus this function.

BTW there is already a similar keyword munging fn, see `database-type` in `upload.clj`.

Am I missing something obvious?

crisptrutski replied:
Fair play I guess, yeah I guess we should be consistent.

I don't think lightweight aliasing was a think when I made that call, but I guess also it felt weird to refer to "submodule" namespaces from outside the module. Maybe the wind has just changed inside me, but I would rather avoided the bidirectional translation, feels the greater of two evils.

Curious about your thoughts, but let's punt on it for now.

wotbrew replied:
Oh yes, totally agree FWIW. We can do a refactor at some point perhaps to have `types.cljc` use the `:metabase.upload` keys. Also agreed now lightweight aliasing exists this is much easier (but even before I would have been guilty of `(create-ns 'metabase.upload) (alias 'metabase.upload 'upload)`)

crisptrutski replied:
Well I really don't want to use the top level namespace for the types, then they're indistinguishable from other kinds of namespaced keywords, like the operation type.

There's a part of me that want to go the other way and have `metabase.upload.value-type` and `metabase.upload.column-type` to avoid where we pun on these keywords, but doing the translations between those two is a bit tedious.

crisptrutski commented on src/metabase/upload.clj:
```diff
@@ -755,48 +763,58 @@
-                                     (alter-columns! driver database table (:updated changes))))
+              modify-schema?      (and (not= old-types new-types) (= detected-types new-types))
+              _                   (when modify-schema?
+                                    (let [changes (field-changes column-names old-types new-types)
+                                          old-types (old-column-types driver column-names old-types)]
```
nit: alignment

wotbrew replied:
a317ceb40e8c

crisptrutski commented on test/metabase/upload_test.clj:
```diff
@@ -2205,28 +2205,64 @@
   (let [round-if-float #(if (float? %) (u/round-to-decimals digits-precision %) %)]
     (mapv (partial mapv round-if-float) rows)))
 
+(defn- external-type [t]
+  (keyword "metabase.upload" (name t)))
```
Ditto on "why force translation?"

crisptrutski reviewed (commented):
The final stretch!

crisptrutski commented on src/metabase/upload/types.cljc:
```diff
@@ -80,13 +80,6 @@
   {::*boolean-int*  ::boolean
    ::*float-or-int* ::float})
 
-;; TODO: the set of allowed promotions should be driver-specific, because not all drivers support coercions between all
-;; types e.g. redshift does not allow coercions except between text types
-(def ^:private allowed-promotions
```
Wow, I had totally forgotten this old method and its suggestion.

How ironic to have argued with my past self 😆 

At least my naming predilections remained the same.

wotbrew committed:
s/upload-promotion-allowlist/allowed-promotions/g

wotbrew committed:
Prefer case over condp

wotbrew committed:
Unlearn boolean/predicate habit

bikeshed: Personal style vs metabase style:
e.g I prefer flags/booleans to not have ?, reserving that only for
questions i.e predicates

wotbrew committed:
Fix alignment

wotbrew committed:
s/internal-allowed-promotions/translate-type-keywords/g

wotbrew committed:
Fix wierd comment line break

wotbrew commented on test/metabase/upload_test.clj:
```diff
@@ -2490,3 +2577,12 @@
+    (testing "driver allow lists should be transitively closed."
+      ;; Imagine it was possible to go from (int -> float) and (bool ->
+      ;; int) - but not (bool -> float). This would mean a user could get (bool -> float), but only by filtering and re-uploading portions of
+      ;; the csv - Yuk! This test ensure drivers always meet any transitive expectations users might have.
```
Good spot 🙏, unfortunately I committed this late with bad eyesight on my tiny laptop screen! 

15ace4135174

wotbrew committed:
Fix missing potemkin import
